lib/elixir/lib/calendar/date.ex

```diff
@@ -757,6 +762,68 @@ defmodule Date do
     end
   end
 
+  @doc """
+  Shifts given `date` by `duration` according to its calendar.
+
+  Allowed units are: `:year, :month, :week, :day`.
+
+  When used with the default calendar `Calendar.ISO`:
+
+  Durations are collapsed before they are applied:
+  - when shifting by 1 year and 2 months the date is actually shifted by 14 months
+  - when shifting by 2 weeks and 3 days the date is shifted by 17 days
+
+  Durations are applied in order of the size of the unit: `month > day`.
+
+  Raises ArgumentError when called with time scale units.
```

Comment:
I can see someone having a Duration with a mix of date-scale and time-scale units wanting to use this function and discarding the time-scale units. E.g. 2 days, 5 minutes, 3 seconds. And you just want to shift by 2 days ignoring the minutes and seconds.

A new function called something like `Duration.truncate_time(duration)`could be called before calling `Date.shift/2`.

Reply:
I am afraid this could cause for some confusion, e.g. if called with a duration of 86400 seconds, the caller might expect that the date got shifted by a month, while the function would have silently done nothing. Since we have a shift function per calendar type, i feel it's ok to expect from the caller to pass a fitting duration depending on which type they're working with.

Reply:
We may add `Duration.truncate_time_units` in the future and `Duration.truncate_date_units` though. It can be the topic of a future pull request.